packages/iov-bcp-types/src/atomicswap.ts

```diff
@@ -0,0 +1,90 @@
+import { Stream } from "xstream";
+
+import { BcpCoin, BcpConnection, BcpQueryEnvelope } from "./bcp";
+import { Address } from "./signables";
+import { SwapIdBytes } from "./transactions";
+
+export enum SwapState {
+  OPEN = "open",
+  CLAIMED = "claimed",
+  EXPIRED = "expired",
+}
+
+export interface SwapData {
+  readonly id: SwapIdBytes; // this is used as an unique identitier to locate the swap
+  readonly sender: Address;
+  readonly recipient: Address;
+  readonly hashlock: Uint8Array; // this is the hash, whose preimage releases the swap
+  readonly amount: ReadonlyArray<BcpCoin>;
+  readonly timeout: number; // blockheight where the swap expires (TODO: alternatively support Date?)
+  readonly memo?: string;
+}
+
+// OpenSwap is an offer that has not yet been claimed
+export interface OpenSwap {
+  readonly kind: SwapState.OPEN;
+  readonly data: SwapData;
+}
+
+// ClosedSwap is returned once the swap has been claimed, exposing the preimage that was used to claim it
+export interface ClaimedSwap {
+  readonly kind: SwapState.CLAIMED;
+  readonly data: SwapData;
+  readonly preimage: Uint8Array;
+}
+
+// ExpiredSwap is an offer that timed out without being claimed
+export interface ExpiredSwap {
+  readonly kind: SwapState.EXPIRED;
+  readonly data: SwapData;
+}
+
+export type BcpAtomicSwap = OpenSwap | ClaimedSwap | ExpiredSwap;
+
+export interface BcpSwapRecipientQuery {
+  readonly recipient: Address;
+}
+
+export interface BcpSwapSenderQuery {
+  readonly sender: Address;
+}
+
+export interface BcpSwapIdQuery {
+  readonly swapid: SwapIdBytes;
+}
+
+// can we implement this as well? maybe id should equal swap
+// export interface SwapHashQuery {
+//   readonly hashlock: Uint8Array;
+// }
+
+export type BcpSwapQuery = BcpSwapRecipientQuery | BcpSwapSenderQuery | BcpSwapIdQuery;
+
+// isQueryBySwapRecipient is a type guard to use in the swap-based queries
+export function isQueryBySwapRecipient(query: BcpSwapQuery): query is BcpSwapRecipientQuery {
+  return (query as BcpSwapRecipientQuery).recipient !== undefined;
```

Comment:
`query` should be cased to `any` to avoid mixing up compile time and runtime properties. Testing an BcpSwapRecipientQuery instance for `.recipient !== undefined` makes no sense: it is always defined.

Is I was a compiler, I would optimize out that check.